Add computation of Jacobian, inverse Jacobian, and determinant of the Jacobian that satisfy the metric identities

nilsdeppe opened the pull request:
## Proposed changes

The function currently in develop only computes `J d_i xi^{\hat{i}}` (det jac times inv jac, not J and inv jac). The new function uses the one in develop but also computes the determinant of the Jacobian, inverse Jacobian, and the Jacobian such that:
 J * inv_jac satisfies metric identities
inv_jac * jac = identity

I split this into:
- a refactor into impl (+164, -130) where the code implementation is unchanged
- a commit that simplifies the `make_map` function in the test by using `if constexpr` (+18, -26)
- a commit that adds the new functionality and tests (+268, -62)

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Factor metric_identity_jacobian into impl

This allows passing a buffer into the metric_identity_jacobian_impl in 3d to
avoid memory allocations. The buffer-passing will be used in a followup commit
that computes the Jacobian, det Jacobian, inverse Jacobian, and det Jacobian
times inverse Jacobian all so that they satisfy the metric identities.

nilsdeppe committed:
Add computation of metric identity Jacobian, etc.

Add a function that computes the Jacobian, inverse Jacobian, determinant of the
Jacobian, and the determinant of the Jacobian times the inverse Jacobian so that
everything satisfies the metric identities and is consistent.

nilsdeppe committed:
Simplify make_map in metric identities via if constexpr

fmahebert commented:
To help my understanding: can you clarify (here in Github) why the first commit is giving the 3d case a special handling with buffer allocations but not the 2d case? does the 2d algorithm simply not need the memory space? or are you not optimizing it as much?

nilsdeppe commented:
The 2d case is just so much simpler than the 3d case (Levi-Civita tensor is a lot easier in 2d :) ) and that's why you don't need it. Basically, in 2d you can get into the situation where you have 2 unused components (and thus 2 buffers) and so taking the eta derivatives you have plenty of space for transposing. In 3d all components of the Jacobian depend on either eta or zeta derivatives and so you can't use one of them as a buffer to do a transpose.

fmahebert commented:
Okay, thanks! I'll try to look at this in the next day or so :)

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.cpp:
```diff
@@ -56,145 +57,178 @@ void metric_identity_jacobian(
 
-    DataVector buffer(mesh.number_of_grid_points());
-    DataVector buffer_component(mesh.number_of_grid_points());
+void metric_identity_jacobian_impl(
```
Can you add a comment to the source code explaining why the 3d impl is split? I think your answer to me, and the commit message, were both useful summaries. Perhaps the same words could be placed in a comment above this line, so someone in the future can get the answer without looking through the code history?

The comment might also be located around line ~218, where the impls are used. I'm not sure. Please do whichever you would find most helpful :)

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.hpp:
```diff
@@ -183,4 +183,46 @@ void metric_identity_jacobian(
+
+/*!
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief Compute the Jacobian. inverse Jacobian, and determinant of the
```
Typo: period -> comma

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.hpp:
```diff
@@ -183,4 +183,46 @@ void metric_identity_jacobian(
+ * Jacobian that satisfies the metric identities.
+ */
+template <size_t Dim>
+void metric_identity_jacobian_and_inv_jacobian(
```
[styling nitpick] Improve match between function name/docs with function args...
1. Reorder arguments to `jacobian`, `inverse_jacobian`, `det_jacobian`, `det_jac_times_inverse_jacobian`
2. [optional] Rename to `metric_identity_jacobian_quantities`

nilsdeppe replied:
1. I don't see why the order you suggest is the preferable one. I would argue that `det_jac_times_inverse_jacobian`, `det_jacobian`, `inverse_jacobian`, `jacobian` is the preferable order because this is the order things are computed in (up to possibly Jacobian & inverse Jacobian). Additionally, having `jacobian` at the end of the list puts it closest to the input arguments which I think would be reasonable given it must be specified on input.
2. sure!

fmahebert replied:
Oh, option 1. was aiming to match the args to the name of the function, but otherwise I agree with you. Plus I see you've done option 2., so I'm satisfied 👍

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.hpp:
```diff
@@ -183,4 +183,46 @@ void metric_identity_jacobian(
+ * \warning on entry `Jacobian` must be the Jacobian to use for compute the
+ * determinant of the Jacobian times the inverse Jacobian so that it satisfies
+ * the metric identities. On output the Jacobian is the inverse of the inverse
+ * Jacobian that satisfies the metric identities.
```
1. Typo: "on entry `Jacobian`" -> "on entry `jacobian`" (lower case) ... because you're describing the argument, not a type.
2. Fix phrasing "to use for compute" (-> "to use when computing"?)
3. Clarify: "On output the Jacobian is" -> "On output `jacobian` is" to clarify you're describing the return argument.

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.hpp:
```diff
@@ -183,4 +183,46 @@ void metric_identity_jacobian(
+ * \brief Compute the Jacobian. inverse Jacobian, and determinant of the
+ * Jacobian so that they satisfy the metric identities.
+ *
+ * Uses `metric_identity_jacobian` to compute the determinant of the Jacobian
```
Now that you're computing more jacobian quantities, I suggest renaming the first function `metric_identity_jacobian` (and its impl) to more accurately reflect what it does. I suggest `metric_identity_det_jac_times_inv_jac(_impl)` or similar (I don't feel strongly which words are shortened vs not 😛). It may be most natural to make this name change in a new commit in this PR, after the commit that does the refactor into the impl?

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.hpp:
```diff
@@ -183,4 +183,46 @@ void metric_identity_jacobian(
+ *     \sqrt[(d-1)]{J^{d-1}} & \mathrm{if} \; d \; \mathrm{even} \\
+ *     \mathrm{sgn}\left(\frac{J^{-1}\partial_\xi x}{\partial_\xi
+ *     x}\right)\sqrt[(d-1)]{J^{d-1}} & \mathrm{if} \; d \; \mathrm{odd}
+ *   \end{array}\right.
```
Haaalp
- Define your symbols
- I expected this would be a formula for detJ in terms of detJinvJ. But it isn't, because the same symbol appears on each side of equal sign... so how is this a useful starting point for computing J/invJ/detJ from detJinvJ?
- Both cases (even + odd) have a common factor (the root of the power of the jacobian) that seems should be just |det(J)|*(±1) for even/odd cases. I think writing it as such would be easier to read.
- The odd case has the sign of (invJ times a term divided by itself). Is this a typo?
- Where does this equations come from? Citation perhaps?

nilsdeppe replied:
I've (significantly) expanded the documentation in a way that hopefully clears up all your questions. There's nothing to cite here (basically the only thing used is a property of determinants, but I think that was not clear before).

fmahebert replied:
Thank you, much clearer now. I've added a few comments in the surrounding text suggesting minor further clarifications.

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.hpp:
```diff
@@ -183,4 +183,46 @@ void metric_identity_jacobian(
+ * \f}
+ *
+ * From this we can compute the inverse Jacobian and then invert it to obtain
+ * the Jacobian that is exactly the inverse of the inverse Jacobian.
```
For my education, can you explain a little (here in Github to start) the process of "exact jac" -> "metric identity inv jac" (I think we're agreeing to use "metric identity X" to mean "X that satisfies the discrete metric identities)-> "metric identity jac" ?
- why do you need the "exact" Jacobian at all... could you not compute the metric identity Jacobian from differentiating the map numerically?
- does it matter what the initial Jacobian is? for instance, if the input Jacobian is already the metric identity Jacobian... do you get the same output? if different output, how do you evaluate the validity/correctness of the different outputs?
- on line 190 above, you've said you start from detJinvJ and compute J/detJ/invJ to satisfy the metric identities. Are there new metric identities for J/detJ/invJ, or do you just mean that these three quantities are consistent with the detJinvJ that satisfies the metric identities?

nilsdeppe replied:
- yes, I've clarified that the input Jacobian (`jacobian`) can be any approximation to the Jacobian. I would like to make this function to able to work for both analytic and numerical Jacobians.
- I think (but haven't verified) that the Jacobian would be changed if you passed in the output of this function. This is attributed to truncation error in compute the Jacobian (and numerical derivatives of it). Probably with enough iterations you'd reach a sort of "equilibrium" value, but all these approximations are equally valid (i.e. differ at or below truncation error). The key thing being that the result always satisfies that the inverse Jacobian and Jacobian are numerical inverse of each other, and the det(J) invJ satisfies the metric identities.
- The latter. I'm hoping my updated documentation makes this clear. If not, please let me know what you think would make it clearer!

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.cpp:
```diff
@@ -232,28 +234,146 @@ void metric_identity_jacobian(
+                                mesh.number_of_grid_points());
+  destructive_resize_components(inverse_jacobian, mesh.number_of_grid_points());
+  destructive_resize_components(det_jacobian, mesh.number_of_grid_points());
+  // ASSERT();
```
Remove... unless if maybe it'd be useful to ASSERT that the input/output `jacobian` also matches the mesh size?

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.cpp:
```diff
@@ -232,28 +234,146 @@ void metric_identity_jacobian(
+    get(*det_jacobian) = get<0, 0>(*jacobian);
+    get<0, 0>(*inverse_jacobian) = 1.0 / get<0, 0>(*jacobian);
+  } else if constexpr (Dim == 2) {
+    // Note that det_jacobian and jacobian are used a buffers here
```
Typo: "used a" -> "used as"

fmahebert reviewed (commented):
A partial review. I'll check the math in the .cpp and the test once I better understand what this is doing :)

nilsdeppe committed:
Rename metric_identity_jacobian

Rename to metric_identity_det_jac_times_inv_jac

nilsdeppe commented:
I've rebased on develop and pushed a fixup commit (and a rename).  There are a couple things left to be decided in github comments before I change them. Thanks for the review!

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.cpp:
```diff
@@ -71,6 +73,15 @@ void metric_identity_jacobian_impl(
+  // depend on either eta or zeta derivatives and so you can't use one of them
+  // as a buffer to do a transpose.
+  //
+  // The 3d implementation is what should be generalized to higher dimensions of
```
Typo: "of" -> "if"

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.hpp:
```diff
@@ -175,12 +175,77 @@ namespace dg {
+ *
+ * Uses `dg::metric_identity_jacobian()` to compute the determinant of the
+ * Jacobian times the inverse Jacobian. Taking the determinant in \f$d\f$
+ * dimensions:
```
I suggest expanding a bit for clarity: "By taking the determinant of this product, we can isolate $J$, the determinant of the Jacobian. In $d$ dimensions, we have:" or similar.

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.hpp:
```diff
@@ -175,12 +175,77 @@ namespace dg {
+ * determinant of the Jacobian times the inverse Jacobian so that it satisfies
+ * the metric identities. On output the `jacobian` is the inverse of the inverse
+ * Jacobian that satisfies the metric identities. The `jacobian` can be computed
+ * analytically or numerically, either is fine.
```
I suggest swapping the order of the last two sentences (i.e., move this last sentence into the middle of the paragraph). I think this will be helpful because it will show that the input can be computed in different ways as soon as you've presented the input.

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.cpp:
```diff
@@ -56,170 +61,337 @@ void metric_identity_jacobian(
+  // Jacobian.
+  if constexpr (Dim == 1) {
+    get(*det_jacobian) = get<0, 0>(*jacobian);
+    get<0, 0>(*inverse_jacobian) = 1.0 / get<0, 0>(*jacobian);
```
Here in 1D you're computing all quantities to be consistent with `jacobian`, not with `det_jac_times_inv_jac`. I guess in 1D it may not matter, but in the spirit of being precise, I suggest adding a comment here pointing out this optimization of sorts.

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.cpp:
```diff
@@ -56,170 +61,337 @@ void metric_identity_jacobian(
+    {
+      // In 3d det(J inv_jac) = 1/J^2.
+      // We need to use a temporary (which we turn into a buffer using the
+      // inverse_jacobian allocation) in order to compute the sign correct.
```
Typo: "the sign correct" -> "the correct sign" OR "the sign correctly"

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.cpp:
```diff
@@ -56,170 +61,337 @@ void metric_identity_jacobian(
+                                    get<1, 1>(*jacobian))}
+                 << " and the (2,2) component gives "
+                 << DataVector{sign(get<2, 2>(jacobian_by_det_jacobian) /
+                                    get<2, 2>(*jacobian))});
```
For my understanding: why are you only checking the diagonal components?

nilsdeppe replied:
That's a great question! I'm worried about round off issues in the off-diagonal components since they could be zero (~1.0e-14) but could flip flop sign conceivably. My worry is some non-trivial map that just happens to have a diagonal (or diagonally very dominant) Jacobian and therefore the off diagonal terms aren't exactly zero.

fmahebert replied:
Isn't there a concern that a map similar to DiscreteRotation could have order-unity off-diagonal components but order-zero diagonal components, making your argument valid but in the other direction?

If this is a concern, perhaps the robust (but obviously somewhat annoying) solution is to check the several largest components?

nilsdeppe replied:
Awe :( Yea, I think it's best to just check the largest 2/3 components. To reduce overhead think it's sufficient to use the `0` grid point to sort by?

fmahebert replied:
I don't have great intuition for this, but I'd tend to agree. Maybe with a note that about the assumptions being made so they can be revisited if needed in the future?

nilsdeppe replied:
Maybe the safest thing to do is to use the sum of one row or column. e.g.
```cpp
(get<0, 0>(jacobian_by_detjacobian) + get<0, 1>(jacobian_by_detjacobian) + get<0, 2>(jacobian_by_detjacobian)) / (...)
```
If an entire row/column is zero then the Jacobian is singular and we are not doing so well :P 

@kidder @fmahebert thoughts?

nilsdeppe replied:
Offline discussion:

The only reason the Jacobian changes sign is because maps that use discrete rotations don't actually rotate, they negate the axes, which causes the coordinate system to go from right-handed to left-handed, effectively. The right solution is to actually fix those coordinate maps and domain creators so that the sign of det(Jacobian) is preserved.

fmahebert commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_MetricIdentityJacobian.cpp:
```diff
@@ -33,81 +33,75 @@ using Affine = domain::CoordinateMaps::Affine;
+      CHECK_ITERABLE_CUSTOM_APPROX(divergence_terms.get(i), expected,
+                                   local_approx);
+    }
   }
```
For my understanding: why did this first set of checks get moved into a scope?

nilsdeppe replied:
If remember correctly it was name collisions, e.g. `jacobian` and `inertial_corods` being reused but in the first case with random values and the second with "motivated" values

fmahebert reviewed (commented):
Good clarify improvements, thank you!

Here are some remaining things. Ping me if I've missed any of your questions to respond to.

nilsdeppe commented:
I've rebased on develop and pushed a fixup. I've changed to assuming that detJ > 0.0 with an ASSERT. It turns out none of the domain creators actually violate this (except the 1d rotated intervals, but I think that's okay to allow). I'll submit a separate PR that adds asserts and fixes the negative detJ in the tests separately. However, this PR doesn't depend on that.

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.cpp:
```diff
@@ -302,63 +314,6 @@ void metric_identity_jacobian_quantities(
     }
-  } else {
-    {
-      // In 3d det(J inv_jac) = 1/J^2.
```
Can you re-add this particular comment line inside the "if constexpr 3d" around line 306? This will I think help clarify the sqrt in 3d.

Edit: tried to better describe suggested location... line numbers change depending on which commit is looked at...

fmahebert reviewed (commented):
Okay this is a lot simpler 👍

One tiny change suggestion, please squash this in directly while you squash to your preferred commits for merging. I'll take another pass at that point (to make sure I didn't get anything confused while looking through the fixups).

nilsdeppe commented:
Okay, rebased, squashed, and added the comment back. Thanks for all the feedback!

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MetricIdentityJacobian.cpp:
```diff
@@ -69,6 +74,15 @@ void metric_identity_jacobian_impl(
+  // as a buffer to do a transpose.
+  //
+  // The 3d implementation is what should be generalized to higher dimensions if
+  // the need arises.
```
This was squashed into the wrong commit (should go in commit that adds impl)

nilsdeppe replied:
:( hopefully done correctly now:P

fmahebert reviewed (approved):
Thanks for addressing all my requests on this!

nilsdeppe commented:
Of course! That's for all the help improving the code :)
